Subject: Allow hotplug on iSeries vio devices
From: Stephen Rothwell <sfr@canb.auug.org.au>                                                                                                                        


Bug 159652 - LTC22358 - iseries lacks a full-featured flat device-tree, no uevents triggered for vio devices


This patch allows "echo add > /sys/bus/vio/devices/vlan0/uevent" to cause the
appripriate module to be loaded.  This works for all vio devices except vscsi. 
Is this sufficient?  A patch to create the device tree nodes for all vio
devices would be much larger (and a bigger risk). 

Signed-off-by: Olaf Hering <olh@suse.de>

---
 arch/powerpc/kernel/vio.c |   15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

Index: linux-2.6.16/arch/powerpc/kernel/vio.c
===================================================================
--- linux-2.6.16.orig/arch/powerpc/kernel/vio.c
+++ linux-2.6.16/arch/powerpc/kernel/vio.c
@@ -267,21 +267,22 @@ static int vio_hotplug(struct device *de
 			char *buffer, int buffer_size)
 {
 	const struct vio_dev *vio_dev = to_vio_dev(dev);
-	char *cp;
+	char *cp = NULL;
 	int length;
 
 	if (!num_envp)
 		return -ENOMEM;
 
-	if (!vio_dev->dev.platform_data)
-		return -ENODEV;
-	cp = (char *)get_property(vio_dev->dev.platform_data, "compatible", &length);
-	if (!cp)
-		return -ENODEV;
+	if (vio_dev->dev.platform_data) {
+		cp = (char *)get_property(vio_dev->dev.platform_data,
+				"compatible", &length);
+		if (!cp)
+			return -ENODEV;
+	}
 
 	envp[0] = buffer;
 	length = scnprintf(buffer, buffer_size, "MODALIAS=vio:T%sS%s",
-				vio_dev->type, cp);
+				vio_dev->type, cp ? cp : "");
 	if (buffer_size - length <= 0)
 		return -ENOMEM;
 	envp[1] = NULL;
